✨ Add custom remediation for workflow permissions/pinned dependencies

laurentsimon opened the pull request:
Add custom remediation for workflow permissions/pinned dependencies
```release-notes
Add custom remediation for workflow permissions/pinned dependencies
```
see https://github.com/ossf/scorecard/issues/1850

Screenshot:
![image](https://user-images.githubusercontent.com/64505099/166841575-573571a8-4a00-431e-a4cd-7cf9fc609063.png)

laurentsimon commented:
/cc @varunsh-coder

laurentsimon commented:
Fyi, the protobuf-related checks have started failing again over the past few days

varunsh-coder commented on checks/remediations.go:
```diff
@@ -0,0 +1,86 @@
+func remdiationSetup(c *checker.CheckRequest) error {
+	remediationOnce.Do(func() {
+		// Get the branch for remediation.
+		b, err := c.RepoClient.GetDefaultBranch()
```
Should this be the default branch or the branch on which Scorecard Action was run? As an example, if one creates a PR with a new workflow, which is missing permissions, this will still have main (/default branch). But when app.stepsecurity.io tries to fix it, it will not find the workflow in that branch.

laurentsimon replied:
You're correct. However we only support the default branch for non pull_request events https://github.com/ossf/scorecard-action/blob/main/entrypoint.sh#L101 and we don't officially support (https://github.com/ossf/scorecard-action below the table of content).

So I think we only need to support the default branch for now. Let me know if I missed some cases, it's possible

varunsh-coder replied:
ok, thanks for the info!

varunsh-coder commented on checks/remediations.go:
```diff
@@ -0,0 +1,86 @@
+		return nil
+	}
+
+	text := fmt.Sprintf("update your workflow using https://app.stepsecurity.io/secureworkflow/%s/%s/%s?enable=%s",
```
Should these literals be set as constants at top of file? Else will need to be modified at multiple places.

laurentsimon replied:
yes I can do that. Thanks!

varunsh-coder commented:
@laurentsimon thanks a lot for the PR! Great to see remediation being added. I added couple of comments. Do you have a screenshot of how this looks right now? I am curious how remediation info will show up in the GitHub Code Scanning UI.

laurentsimon commented:
> @laurentsimon thanks a lot for the PR! Great to see remediation being added. I added couple of comments. Do you have a screenshot of how this looks right now? I am curious how remediation info will show up in the GitHub Code Scanning UI.

I added the screenshot i the issue PR description. PTAL

azeemshaikh38 commented on checks/remediations.go:
```diff
@@ -0,0 +1,89 @@
+	return remediationSetupErr
+}
+
+func createWorkflowPermissionRemediation(filepath string) *checker.Remediation {
+	return createWorkflowRemediation(filepath, "permissions")
+}
+
+func createWorkflowPinningRemediation(filepath string) *checker.Remediation {
+	return createWorkflowRemediation(filepath, "pin")
+}
+
+func createWorkflowRemediation(path, t string) *checker.Remediation {
```
Let's pass the repo name and branch as arguments to these fns directly. The checks already take care of ensuring only supported repo types call them so it'll avoid all the extra complexity of checking it here again.
